src/main/java/kafka/Consumer.java

```diff
@@ -3,47 +3,38 @@
 import configuration.Config;
 import java.time.Duration;
 import java.util.Date;
-import java.util.stream.StreamSupport;
 import monitoring.Monitor;
-import org.apache.kafka.clients.consumer.ConsumerRecord;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 import reactor.core.scheduler.Schedulers;
+import reactor.kafka.receiver.KafkaReceiver;
 import target.ITarget;
 
 public class Consumer {
 
-    private final ReactiveKafkaClient<String, String> kafkaConsumer;
+    private final KafkaReceiver<String, String> kafkaReceiver;
     private final ITarget target;
 
-    public Consumer(ReactiveKafkaClient<String, String> kafkaConsumer, ITarget target) {
-        this.kafkaConsumer = kafkaConsumer;
+    public Consumer(KafkaReceiver<String, String> kafkaReceiver, ITarget target) {
+        this.kafkaReceiver = kafkaReceiver;
         this.target = target;
     }
 
     public Flux<?> stream() {
-        return kafkaConsumer
-            .doOnNext(records -> Monitor.batchProcessStarted(records.count()))
+        return kafkaReceiver
+            .receiveBatch()
+            .flatMap(records -> records.count().doOnNext(Monitor::batchProcessStarted).thenReturn(records))
             .concatMap(records -> {
                 var batchStartTimestamp = new Date().getTime();
-                var hasNullKey = StreamSupport
-                    .stream(records.spliterator(), false)
-                    .anyMatch(record -> record.key() == null);
-
-                return Flux
-                    .fromIterable(records)
-                    .groupBy(record -> hasNullKey ? record.partition() : record.key())
+                return records
+                    .groupBy(x -> x.key() == null ? x.partition() : x.key())
                     .delayElements(Duration.ofMillis(Config.PROCESSING_DELAY))
                     .publishOn(Schedulers.parallel())
                     .flatMap(partition -> partition.concatMap(record -> Mono.fromFuture(target.call(record))))
                     .collectList()
-                    .map(__ -> batchStartTimestamp);
-            })
-            .doOnNext(Monitor::batchProcessCompleted)
-            .map(__ -> {
-                kafkaConsumer.commit();
-                return 0;
+                    .doOnNext(__ -> Monitor.batchProcessCompleted(batchStartTimestamp))
+                    .map(__ -> records);
             })
-            .doOnNext(__ -> kafkaConsumer.poll());
+            .flatMap(records -> records.last().flatMap(record -> record.receiverOffset().commit()));
```

Comment:
Reacto kafka provides multiple methods to commit, I'm not sure of the implication of doing it ourselves or letting the library do it, it has for example:
- `receiveAutoAck` instead of `receiveBatch`
- or `record.receiverOffset().acknowledge()` and the library will commit itself

Reply:
we want to consume a batch, process it, and make sure we commit the batch before we process the next one. 
that was the idea of the implementation before, but I think it was implemented in a buggy-potential way.